### ISSUE TITLE
Joomla content in Particles

### ISSUE THREAD

**Issue by JoomFX** [gantry/gantry5#1225]
Hey Matias,

We have discussed this several times already, but I'm opening this issue for two reasons.
First, to have it as a reminder and second, to forward all users that ask about it here.

I have several users asking why they cannot show Joomla articles in the particles.

So, could you please explain the whole thing once and forever. Just few sentences explaining why it is currently not possible, what are the future plans (Web Services in Joomla), etc.

Also, it seems that [Web Services will not make it for Joomla 3.6](https://groups.google.com/forum/#!msg/joomla-dev-cms/ALeU_2_SKAo/YtUwzZ98AAAJ), it seems that it will be postponed for 3.7 and with the current release cycle that Joomla has, it means at the end of 2016 or in the Spring of 2017. If this is the case, do you have plans to implement this feature in Gantry 5 itself, without the need of Web Services? Thanks :wink:

**Comment by mahagr** [comment]
Other platforms like WP and Grav provide API to get post or page data from the system. So loading an article in those platforms is as simple as making a function call with the id to the wanted article.

In Joomla there is no function call to be made; each extension needs to write code to fetch the article data from the database. So basically each view in Articles component has its own logic as do all the modules in helper.php file (archive, categories, category, latest, new and popular...) where they basically do the same as the component -- fetch the data directly from the database.

This means that Joomla doesn't have API for articles and everyone who wants to load the data, need to write their own custom logic to do that. Basically that means a lot of copying of the code and hoping that Joomla never changes its database.

This is really, really bad design and its maybe the worst architectural flaw in Joomla and most of its extensions.

I was hoping to see API in J3.6, but I guess that will not happen yet. Alright, so it means that I have to do it all myself, lol.

PS: Same question was asked in #1160. :)

**Comment by JoomFX** [comment]
Thanks a lot for the reply Matias!

By the way, you might not like it, but I'm really happy that **you** will be implementing it (rather then using a Joomla feature) because that way I'll know it is implemented properly and will work naturally with G5 :wink: 

I'm looking forward to see this implemented.
That's how your baby (the particles) will become something really, really BIG!

P.S.  Sorry for opening a duplicate issue, I didn't know about https://github.com/gantry/gantry5/issues/1160 :wink:

**Comment by mahagr** [comment]
I have added basic support to fetch Joomla articles and categories into particles. No work has been done to enable the same in admin side yet.

Here is example of a particle which fetches Joomla sample categories (I think I used test sample) and featured articles.

```
{% extends '@nucleus/partials/particle.html.twig' %}

{% block particle %}
    {% set categories = joomla.finder('category').language().published().id(14, 3).find() %}

    {% for category in categories %}
        <h2><a href="{{ category.route }}">{{ category.title }}</a> (id {{ category.id }}) <small>{{ category.path }}</small></h2>
        {% if category.params.image %}
            <img src="{{ url(category.params.image) }}" style="border: 5px solid black" />
        {% endif %}
        {{ category.description|default('<p>No Description.</p>')|raw }}
        <div class="alert alert-info">{{ category.note }}</div>
    {% endfor %}


    {% set latest = joomla.finder('article').category(categories).language().featured().order('id', -1).limit(3).find() %}

    {% for article in latest %}
        <div>
            <h2><a href="{{ article.route }}">{{ article.title }}</a> <small>Id: {{ article.id }}</small></h2>
            <div>By {{ article.created_by_alias }} on {{ article.created }}</div>
            <div>
                Category:
                {% for category in article.categories|reverse %}
                    <a href="{{ category.route }}">{{ category.title }}</a> /
                {% endfor %}
            </div>

            <div class="alert alert-info">{{ article.introtext|raw }}</div>
            <div>{{ article.fulltext|raw }}</div>

            <div>{{ article.images|json_encode }}</div>
            <div>{{ article.urls|json_encode }}</div>
            <div>{{ article.attribs|json_encode }}</div>
            <div>{{ article.metadata|json_encode }}</div>
        </div>
    {% endfor %}
{% endblock %}
```

I would love to get some help here; I want to have some basic content particles into Gantry itself.

**Comment by JoomFX** [comment]
Thanks a lot for that Matias!
I see that you have written quite some code and I really, really appreciate that.

> No work has been done to enable the same in admin side yet.

What do you mean by that? I hope that it means that there will be an input type (similar to `select.select`) that will automatically list the names of the Categories.
Or will it work with manually typing the Category/Article ID in an `input.text` field for example?

I haven't yet tested the above (and the CI build), I'll do that tomorrow.
Meanwhile, could you please break down the following lines a little bit, with just some basic explanation so everybody can find their way:

```
{% set categories = joomla.finder('category').language().published().id(14, 3).find() %}
```

```
{% set latest = joomla.finder('article').category(categories).language().featured().order('id', -1).limit(3).find() %}
```

> I would love to get some help here; I want to have some basic content particles into Gantry itself.

Let me know if I can help or if this message is meant for Djamil :wink:

**Comment by JoomFX** [comment]
Hey Matias,

I couldn't wait for tomorrow so I installed the CI build and just played with the new feature.
It is really awesome and works perfectly :smile: 

There are just few things that I still do not quite understand and therefore I'd like to ask you the following questions:

1. Using your example code above, the `{{ article.created_by_alias }}` was returning an empty string. I realized that it is normal because the "Created by Alias" field was empty in the Joomla Admin. It is not a common practice to fill out that field, it is usually empty. When I filled out the field, the value was properly shown on the frontend. 
Then I tried `{{ article.created_by }}` and it returned the correct user, but it returned the ID of that user, not its name, for example "Super User". How can I fetch the name of the user from the "Created By" field?

2. `{{ article.images|json_encode }}`  returns the JSON representation as it is supposed to. But how can I get only the image path which I can then put in an `<img>` tag in order to load the image on the front-end? I tried with `{{ url(article.params.image_intro) }}` but it doesn't work.

3. Why do you use [reverse](http://twig.sensiolabs.org/doc/filters/reverse.html) in your example code for fetching the category? I removed that filter and it works the same way:

```
            <div>
                Category:
                {% for category in article.categories|reverse %}
                    <a href="{{ category.route }}">{{ category.title }}</a> /
                {% endfor %}
            </div>
```

4. `{{ article.introtext|raw }}` and `{{ article.fulltext|raw }}` work perfectly fine, but is there another expression that will show the whole article text all together (Introtext and Fulltext together)? This is really needed because if you want to show just the first 100 words of the article and you do not know if the article has Introtext or not, you simply do not know which of the two to use.

5. In the following code, `14` is the Category ID, but what is the `3` for?

```
{% set categories = joomla.finder('category').language().published().id(14, 3).find() %}
```

6. The above code sets the category from which to fetch articles. How can I set 2 or more categories to fetch articles from at the same time?

Basically, that's all :wink: 
I managed to sort (both ascending and descending) by `hits` and `created`, which is awesome :smile: 
I also figured out that we can use `{{ article.hits }}` which is not included in your example.

Matias, man.... thanks a million for this ;)

**Comment by mahagr** [comment]
Cannot sleep either, lol.

1. You can use `article.author.name` (author is currently JUser, though I'm likely going to change that)
2. You got wrong variable, there's no params in article: `{{ url(article.images.image_intro) }}`
3. It returns full list of categories in reverse order (from article to root), I may want to change that... Dunno.
4. Its coming, the base classes I used are a bit old and restricted (from Kunena)
5. `id(14, 3)` means something like: take article 14 and all its children limiting to 3rd level
5. Use array: `id([14,15,16,17], 3)`

**Comment by JoomFX** [comment]
Can't believe you are still awake :smile: 

Everything works perfectly fine now - author name (Created By) was shown properly, the image is shown... everything just works!

Matias, do you realize how many doors you just opened with this feature, do you? :smile: 

Regarding the Admin stuff (YAML) - are you planning to have an input field that will show the **names** of all available Categories (will require a DB call to take the names) or we will just use it with a `input.text` field where the user will type in the Category ID manually? 
Hmm, now when I think about it, it might be better to leave it with manual IDs, because this way the user will be able to add multiple categories, separated by commas. It is up to you, I'm already happy as hell :wink: 

Will wait for an update regarding the `reverse` thing (if you will change it) and for the full article text all together.

Go to bed now, and thanks again!

**Comment by JoomFX** [comment]
Matias, just a quick question..
The following works fine for limiting the text:

`{{ article.fulltext|raw|striptags|slice(0, 15) }}`

but [truncate](http://twig.sensiolabs.org/doc/extensions/text.html) doesn't work, it breaks the particle:

`{{ article.fulltext|raw|striptags| truncate(15) }}`

Truncate is a little bit nicer because it adds `...` at the end, which informs the user that there is more to it.

It seems that Twig Extensions Library (required by `truncate`) is not included in Gantry 5. If this is the case, do you have any plans of including it?

**Comment by mahagr** [comment]
Added support for simple admin parameters and fixed the reverse ordering of categories.

Next I need to look how to improve collections of categories/articles and work out some admin fields.

Here are some docs:

== All finders

=== Creating a new finder

```twig
{% set options = {published: true, id: [14, 3]} %}
{% set finder = joomla.finder('category', options) %}
```
=== `.parse(options)` parse options array to rules

Does the same as the second parameter when creating the finder.

```twig
{% set options = {published: true, id: [14, 3]} %}
{% do finder.parse(options) %}
```

=== `.start(limitstart)` Pagination start

```twig
{% do finder.start(0) %}
```

=== `.limit(limit)` Pagination limit (number of items)

```twig
{% do finder.limit(10) %}
```
=== `.order(by, direction, alias)` Pagination order (order of items)

```twig
{% do finder.start('id', 'asc') %}
```

```twig
{% do finder.limit(10) %}
```
=== `.where(field, operation, value)` Custom where clause

Operation is one of: `>`, `>=`, `<`, `<=`, `=`, `BETWEEN`,  `IN` and `NOT IN`.

```twig
{% do finder.where('id', 'IN', [1,2,3,4,5]) %}
```

**Comment by nicutor** [comment]
Great work @mahagr
